File: sdks/python/pmxt/escrow.py
"""Hosted escrow transaction builders for PMXT clients."""

from __future__ import annotations

import math
from decimal import Decimal, InvalidOperation
from typing import TYPE_CHECKING, Any, Literal, cast
from urllib.parse import quote

from ._hosted_errors import MissingWalletAddress
from ._hosted_routing import _trading_request
from .errors import ValidationError

if TYPE_CHECKING:
    from .client import Exchange


_APPROVAL_TOKENS = frozenset({"usdc", "ctf"})
_WITHDRAW_ACTIONS = frozenset({"request", "claim", "cancel"})
_USDC_SCALE = Decimal("1000000")


def _is_decimal_token_id(token: str) -> bool:
    return token.isascii() and token.isdecimal()


def _approval_token(token: str) -> str:
    if not isinstance(token, str):
        raise ValidationError(
            "token must be 'usdc', 'ctf', or a decimal CTF token_id string",
            field="token",
        )

    candidate = token.strip()
    normalized = candidate.lower()
    if normalized in _APPROVAL_TOKENS:
        return normalized
    if _is_decimal_token_id(candidate):
        return candidate
    raise ValidationError(
        "token must be 'usdc', 'ctf', or a decimal CTF token_id string",
        field="token",
    )


def _amount_wei(value: int | None) -> int | None:
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValidationError(
            "amount_wei must be a non-negative integer",
            field="amount_wei",
        )
    if value < 0:
        raise ValidationError("amount_wei must be non-negative", field="amount_wei")
    return value


def _usdc_amount(value: float | Decimal, *, field: str = "amount") -> float:
    try:
        amount = Decimal(str(value))
    except (InvalidOperation, ValueError) as exc:
        raise ValidationError(
            f"{field} must be a finite positive number",
            field=field,
        ) from exc

    if not amount.is_finite():
        raise ValidationError(f"{field} must be a finite positive number", field=field)
    if amount <= 0:
        raise ValidationError(f"{field} must be positive", field=field)

    scaled = amount * _USDC_SCALE
    if scaled != scaled.to_integral_value():
        raise ValidationError(
            f"{field} precision exceeds 6 decimals; max precision for USDC is 0.000001",
            field=field,
        )

    json_amount = float(amount)
    if not math.isfinite(json_amount):
        raise ValidationError(f"{field} is too large to encode as JSON", field=field)
    return json_amount


class Escrow:
    """Hosted escrow namespace attached to hosted exchange clients."""

    __slots__ = ("_client",)

    def __init__(self, client: Exchange) -> None:
        self._client = client

    def _wallet_address(self, address: str | None = None) -> str:
        resolved = address or getattr(self._client, "wallet_address", None)
        if not resolved:
            raise MissingWalletAddress(
                "wallet_address is required; pass wallet_address to the "
                "Exchange constructor"
            )
        return str(resolved)

    def _request(
        self,
        method: Literal["GET", "POST"],
        path: str,
        body: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return cast(
            dict[str, Any],
            _trading_request(self._client, method=method, path=path, body=body),
        )

    def approve_tx(self, token: str, amount_wei: int | None = None) -> dict[str, Any]:
        """Build an unsigned approval transaction for USDC or CTF escrow access."""
        address = self._wallet_address()
        approval_amount = _amount_wei(amount_wei)
        body = {
            "token": _approval_token(token),
            "user_address": address,
            **({} if approval_amount is None else {"amount_wei": approval_amount}),
        }
        return self._request("POST", "/v0/escrow/approve", body)

    def deposit_tx(self, amount: float | Decimal) -> dict[str, Any]:
        """Build an unsigned USDC deposit transaction into PreFundedEscrow."""
        body = {
            "token": "usdc",
            "amount": _usdc_amount(amount),
            "user_address": self._wallet_address(),
        }
        return self._request("POST", "/v0/escrow/deposit", body)

    def withdraw_tx(
        self,
        action: Literal["request", "claim", "cancel"],
        amount: float | None = None,
    ) -> dict[str, Any]:
        """Build an unsigned USDC withdrawal request, claim, or cancel transaction."""
        if not isinstance(action, str) or action not in _WITHDRAW_ACTIONS:
            raise ValidationError(
                "action must be 'request', 'claim', or 'cancel'",
                field="action",
            )

        address = self._wallet_address()
        if action == "request":
            if amount is None:
                raise ValidationError(
                    "amount is required when action='request'",
                    field="amount",
                )
            body = {
                "action": action,
                "token": "usdc",
                "amount": _usdc_amount(amount),
                "user_address": address,
            }
            return self._request("POST", "/v0/escrow/withdraw", body)

        if amount is not None:
            raise ValidationError(
                f"amount must be omitted when action='{action}'",
                field="amount",
            )

        body = {"action": action, "token": "usdc", "user_address": address}
        return self._request("POST", "/v0/escrow/withdraw", body)

    def withdrawals(
        self,
        include: str = "pending,events",
        address: str | None = None,
    ) -> dict[str, Any]:
        """Return pending withdrawal state and/or withdrawal events for an address."""
        include_value = include.strip()
        if not include_value:
            raise ValidationError("include must not be empty", field="include")

        wallet_address = quote(self._wallet_address(address), safe="")
        include_query = quote(include_value, safe=",")
        path = f"/v0/escrow/{wallet_address}/withdrawals?include={include_query}"
        return self._request("GET", path)


__all__ = ["Escrow"]

File: sdks/typescript/pmxt/escrow.ts
/**
 * Hosted-mode escrow namespace.
 *
 * Exposes the `/v0/escrow/*` endpoints of the pmxt hosted trading API as a
 * small ergonomic helper class. Each method returns the raw upstream JSON
 * payload as `unknown` — typed mappers can be layered on later without
 * breaking the wire shape.
 *
 * Mirrors `sdks/python/pmxt/escrow.py`.
 */

import {
    _tradingRequest,
    formatRoutePath,
    HOSTED_METHOD_ROUTES,
    HostedClientLike,
    resolveWalletAddress,
} from "./hosted-routing";
import { ValidationError } from "./errors";

const APPROVAL_TOKENS = new Set(["usdc", "ctf"]);
const WITHDRAW_ACTIONS = new Set(["request", "claim", "cancel"]);
const USDC_SCALE = 1_000_000n;

function isDecimalTokenId(token: string): boolean {
    return /^[0-9]+$/.test(token);
}

function normalizeApprovalToken(token: string): string {
    if (typeof token !== "string") {
        throw new ValidationError(
            "token must be 'usdc', 'ctf', or a decimal CTF token_id string",
            "token",
        );
    }

    const candidate = token.trim();
    const normalized = candidate.toLowerCase();
    if (APPROVAL_TOKENS.has(normalized)) {
        return normalized;
    }
    if (isDecimalTokenId(candidate)) {
        return candidate;
    }
    throw new ValidationError(
        "token must be 'usdc', 'ctf', or a decimal CTF token_id string",
        "token",
    );
}

function normalizeAmountWei(amountWei: bigint | undefined): string | undefined {
    if (amountWei === undefined) {
        return undefined;
    }
    if (typeof amountWei !== "bigint") {
        throw new ValidationError("amount_wei must be a non-negative integer", "amount_wei");
    }
    if (amountWei < 0n) {
        throw new ValidationError("amount_wei must be non-negative", "amount_wei");
    }
    return amountWei.toString();
}

function validateUsdcDecimal(value: string, field: string): void {
    const match = /^([0-9]+)(?:\.([0-9]+))?$/.exec(value);
    if (!match) {
        throw new ValidationError(`${field} must be a finite positive number`, field);
    }

    const fractional = match[2] ?? "";
    if (fractional.length > 6) {
        throw new ValidationError(
            `${field} precision exceeds 6 decimals; max precision for USDC is 0.000001`,
            field,
        );
    }

    const scaled = BigInt(match[1]) * USDC_SCALE + BigInt(fractional.padEnd(6, "0"));
    if (scaled <= 0n) {
        throw new ValidationError(`${field} must be positive`, field);
    }
}

function normalizeUsdcAmount(value: number | string | bigint, field: string = "amount"): number | string {
    if (typeof value === "bigint") {
        if (value <= 0n) {
            throw new ValidationError(`${field} must be positive`, field);
        }
        return value.toString();
    }
    if (typeof value === "number") {
        if (!Number.isFinite(value)) {
            throw new ValidationError(`${field} must be a finite positive number`, field);
        }
        validateUsdcDecimal(String(value), field);
        return value;
    }
    if (typeof value === "string") {
        const trimmed = value.trim();
        validateUsdcDecimal(trimmed, field);
        return trimmed;
    }
    throw new ValidationError(`${field} must be a finite positive number`, field);
}

function normalizeWithdrawAction(action: string): "request" | "claim" | "cancel" {
    if (typeof action === "string" && WITHDRAW_ACTIONS.has(action)) {
        return action as "request" | "claim" | "cancel";
    }
    throw new ValidationError("action must be 'request', 'claim', or 'cancel'", "action");
}

export class Escrow {
    constructor(private readonly client: HostedClientLike) {}

    /**
     * Build an unsigned approve transaction for a given ERC-20 `token`. When
     * `amountWei` is omitted, the server returns an unlimited approval.
     */
    async approveTx(token: string, amountWei?: bigint): Promise<unknown> {
        const address = resolveWalletAddress(this.client);
        const approvalAmount = normalizeAmountWei(amountWei);
        const route = HOSTED_METHOD_ROUTES.get("escrowApproveTx")!;
        return _tradingRequest(this.client, {
            method: route.method,
            path: route.path,
            body: {
                token: normalizeApprovalToken(token),
                user_address: address,
                ...(approvalAmount === undefined ? {} : { amount_wei: approvalAmount }),
            },
        });
    }

    /**
     * Build an unsigned deposit transaction for `amount` (USDC, 6-decimal
     * grid). Accepts number, decimal string, or BigInt in micro-units.
     */
    async depositTx(amount: number | string | bigint): Promise<unknown> {
        const address = resolveWalletAddress(this.client);
        const route = HOSTED_METHOD_ROUTES.get("escrowDepositTx")!;
        return _tradingRequest(this.client, {
            method: route.method,
            path: route.path,
            body: {
                token: "usdc",
                amount: normalizeUsdcAmount(amount),
                user_address: address,
            },
        });
    }

    /**
     * Build an unsigned withdraw transaction. `action` selects the stage of
     * the withdrawal lifecycle: `request` initiates, `claim` finalizes after
     * the timelock, `cancel` aborts a pending request.
     */
    async withdrawTx(
        action: "request" | "claim" | "cancel",
        amount?: number | string | bigint,
    ): Promise<unknown> {
        const normalizedAction = normalizeWithdrawAction(action);
        const address = resolveWalletAddress(this.client);
        const route = HOSTED_METHOD_ROUTES.get("escrowWithdrawTx")!;
        if (normalizedAction === "request") {
            if (amount === undefined) {
                throw new ValidationError("amount is required when action='request'", "amount");
            }
            return _tradingRequest(this.client, {
                method: route.method,
                path: route.path,
                body: {
                    action: normalizedAction,
                    token: "usdc",
                    amount: normalizeUsdcAmount(amount),
                    user_address: address,
                },
            });
        }
        if (amount !== undefined) {
            throw new ValidationError(
                `amount must be omitted when action='${normalizedAction}'`,
                "amount",
            );
        }
        return _tradingRequest(this.client, {
            method: route.method,
            path: route.path,
            body: { action: normalizedAction, token: "usdc", user_address: address },
        });
    }

    /**
     * List the user's withdrawal records. `include` is forwarded verbatim and
     * defaults to `"pending,events"` to match the Python SDK.
     */
    async withdrawals(
        opts: { include?: string; address?: string } = {},
    ): Promise<unknown> {
        const address = resolveWalletAddress(this.client, opts.address);
        const include = (opts.include ?? "pending,events").trim();
        if (!include) {
            throw new ValidationError("include must not be empty", "include");
        }
        const route = HOSTED_METHOD_ROUTES.get("escrowWithdrawals")!;
        const path = formatRoutePath(route, { address });
        return _tradingRequest(this.client, {
            method: route.method,
            path,
            params: { include },
        });
    }
}
